### PR TITLE
CHK-1656 tokenized email for carts (redis cache) 

### PR DESCRIPTION
**Pull request by infantesimone**
#### List of Changes

- import from https://github.com/pagopa/pagopa-ecommerce-commons logic to tokenize an email (waiting for lib integration)
- update env value with pdv configuration
- tokenize an email (with pdv) to save on redis cache
- update tests

#### Motivation and Context

The goal of PR is to save the tokenized email with Personal Data Vault (pdv) in the cart (so in redis cache).
The business logic for invoking odv is copy&paste from https://github.com/pagopa/pagopa-ecommerce-commons and that's why PR introduces many lines of code ( but this is useful for the future integration with the library)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,20 +7,22 @@ with redirects to [_pagoPA – Checkout_](https://checkout.pagopa.it).
...
+| NODE_FOR_PSP_URI                   | Nodo for PSP URI                                                             | string        |         |
+| NODO_READ_TIMEOUT                  | Http read timeout for all call made to Nodo                                  | number        |         |
+| NODO_CONNECTION_TIMEOUT            | Http connection timeout for all call made to Nodo                            | number        |         |
+| NODO_CONNECTION_STRING             | Connection string containing information used to make Nodo calls             | json (string) |         |
+| CARTS_MAX_ALLOWED_PAYMENT_NOTICES  | Max allowed number of payment notices to be processed for a POST carts request | number        |         |
+| PERSONAL_DATA_VAULT_API_KEY        | API Key for Personal Data Vault (PDV is used to safely encrypt PIIs, e.g. the user's email address)                  | string  |         |
+| PERSONAL_DATA_VAULT_API_BASE_PATH  | API base path for Personal Data Vault                                                                                | string  |         |
```
Comment (giovanniberti) [inline README.md:25]: Reformat:
```suggestion
| Variable name                     | Description                                                                                         | type          | default |
|-----------------------------------|-----------------------------------------------------------------------------------------------------|---------------|---------|
| CHECKOUT_URL                      | Redirection URL for Checkout carts                                                                  | url (string)  |         |
| REDIS_HOST                        | Host where the redis instance used to persist idempotency keys can be found                         | string        |         |
| REDIS_PASSWORD                    | Password used for connecting to Redis instance                                                      | string        |         |
| REDIS_PORT                        | Port used for connecting to Redis instance                                                          | number        |         |
| REDIS_SSL_ENABLED                 | Whether SSL is enabled when connecting to Redis                                                     | boolean       |         |
| NODO_HOSTNAME                     | Nodo connection host name                                                                           | string        |         |
| NODO_PER_PSP_URI                  | Nodo per PSP URI                                                                                    | string        |         |
| NODE_FOR_PSP_URI                  | Nodo for PSP URI                                                                                    | string        |         |
| NODO_READ_TIMEOUT                 | Http read timeout for all call made to Nodo                                                         | number        |         |
| NODO_CONNECTION_TIMEOUT           | Http connection timeout for all call made to Nodo                                                   | number        |         |
| NODO_CONNECTION_STRING            | Connection string containing information used to make Nodo calls                                    | json (string) |         |
| CARTS_MAX_ALLOWED_PAYMENT_NOTICES | Max allowed number of payment notices to be processed for a POST carts request                      | number        |         |
| PERSONAL_DATA_VAULT_API_KEY       | API Key for Personal Data Vault (PDV is used to safely encrypt PIIs, e.g. the user's email address) | string        |         |
| PERSONAL_DATA_VAULT_API_BASE_PATH | API base path for Personal Data Vault                                                               | string        |         |
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/confidential/ConfidentialConfig.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+
+@Configuration
+class ConfidentialConfig {
+
+  @Bean
+  fun personalDataVaultApiClient(
+    @Value("\${confidentialDataManager.personalDataVault.apiKey}") personalDataVaultApiKey: String?,
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/confidential/ConfidentialConfig.kt:15]: ### 
```suggestion
    @Value("\${confidentialDataManager.personalDataVault.apiKey}") personalDataVaultApiKey: String,
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/confidential/ConfidentialConfig.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    @Value("\${confidentialDataManager.personalDataVault.apiBasePath}") apiBasePath: String?
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/confidential/ConfidentialConfig.kt:16]:
```suggestion
    @Value("\${confidentialDataManager.personalDataVault.apiBasePath}") apiBasePath: String
```

File: src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/confidential/ConfidentialConfig.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  ): TokenApi {
+    val pdvApiClient = ApiClient()
+    pdvApiClient.setApiKey(personalDataVaultApiKey)
+    pdvApiClient.basePath = apiBasePath
+    return TokenApi(pdvApiClient)
+  }
+
+  @Bean
+  fun emailConfidentialDataManager(personalDataVaultApi: TokenApi?): ConfidentialDataManager {
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/payment/requests/configurations/confidential/ConfidentialConfig.kt:25]:
```suggestion
  fun emailConfidentialDataManager(personalDataVaultApi: TokenApi): ConfidentialDataManager {
```

File: helm/values-dev.yaml [hunk 5]
```diff
@@ -85,10 +85,12 @@ microservice-chart:
     ELASTIC_APM_LOG_SENDING: "false"
     ECS_SERVICE_NAME: pagopa-ecommerce-payment-requests-service
     ECS_SERVICE_ENVIRONMENT: "dev"
+    PERSONAL_DATA_VAULT_API_BASE_PATH: "https://api.uat.tokenizer.pdv.pagopa.it/tokenizer/v1"
```
Comment (ciuffagianluca) [inline helm/values-dev.yaml:88]: Is it a typo or we use uat pdv in dev?

Comment (infantesimone) [reply helm/values-dev.yaml:88]: use the same pdv test environment (api.uat.tokenizer.pdv.pagopa.it) for dev and uat (so api.dev.platform.pagopa.it and api.uat.platform.pagopa.it)

**Review by ciuffagianluca — APPROVED** [review]
Great job! 👍

File: src/test/kotlin/it/pagopa/ecommerce/payment/requests/utils/TokenizerEmailUtilsTests.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    /* preconditions */ Mockito.`when`(confidentialDataManager.encrypt(email))
+      .thenReturn(Mono.just(computedConfidential))
+    Mockito.`when`(
+        confidentialDataManager.decrypt(
+          ArgumentMatchers.eq(computedConfidential), ArgumentMatchers.any()))
+      .thenReturn(Mono.just(email))
+
+    /* test */
+    val encrypted: Confidential<Email> = tokenizerEmailUtils.toConfidential(email).block()!!
```
Comment (Fernando-Granato) [inline src/test/kotlin/it/pagopa/ecommerce/payment/requests/utils/TokenizerEmailUtilsTests.kt:36]: Same above 🙂
```suggestion
    StepVerifier.create(tokenizerEmailUtils.toConfidential)
          .consumeNextWith(next -> {"assert to execute"})
          .verifyComplete();`
   ```

File: src/test/kotlin/it/pagopa/ecommerce/payment/requests/utils/TokenizerEmailUtilsTests.kt [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    val decrypted: Email = tokenizerEmailUtils.toEmail(encrypted).block()!!
```
Comment (Fernando-Granato) [inline src/test/kotlin/it/pagopa/ecommerce/payment/requests/utils/TokenizerEmailUtilsTests.kt:37]: Same above 🙂

File: src/test/java/it/pagopa/ecommerce/payment/requests/utils/confidential/ConfidentialTest.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        Confidential<Email> confidentialEmail = this.confidentialDataManager.encrypt(email)
+                .block();
+
+        String serialized = objectMapper.writeValueAsString(confidentialEmail);
+
+        TypeReference<Confidential<Email>> typeRef = new TypeReference<>() {
+        };
+        Confidential<Email> deserialized = objectMapper.readValue(serialized, typeRef);
+
+        Email decryptedEmail = confidentialDataManager.decrypt(deserialized, Email::new).block();
```
Comment (Fernando-Granato) [inline src/test/java/it/pagopa/ecommerce/payment/requests/utils/confidential/ConfidentialTest.java:81]: Same above 🙂
```suggestion
         StepVerifier.create(confidentialDataManager.decrypt)
          .consumeNextWith(next -> {"assert to execute"})
          .verifyComplete();`
```

File: src/test/java/it/pagopa/ecommerce/payment/requests/utils/confidential/ConfidentialTest.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    void confidentialJsonRepresentationIsOK() throws JsonProcessingException {
+        Email email = new Email("foo@example.com");
+
+        /* preconditions */
+        Mockito.when(personalDataVaultClient.saveUsingPUT(any()))
+                .thenReturn(Mono.just(new TokenResourceDto().token(UUID.randomUUID())));
+
+        /* test */
+        Confidential<Email> confidentialEmail = this.confidentialDataManager.encrypt(email)
```
Comment (Fernando-Granato) [inline src/test/java/it/pagopa/ecommerce/payment/requests/utils/confidential/ConfidentialTest.java:44]: You can use StepVerifier and not perform a blocking operation
```suggestion
        StepVerifier.create(confidentialDataManager.decrypt)
          .consumeNextWith(next -> {"assert to execute"})
          .verifyComplete();`
```

**Pull request merged**